libs/ballot-interpreter-nh-next/src/geometry.rs

```diff
@@ -0,0 +1,439 @@
+        let dy = self.end.y - self.start.y;
+        Point::new(dx, dy)
+    }
+}
+
+/// Determines an intersection point of two line segments. If `bounded` is set
+/// to `true`, the intersection point must be within the bounds of both
+/// segments. If `bounded` is set to `false`, the intersection point may be
+/// outside the bounds of either segment.
+pub fn intersection_of_lines(
+    segment1: &Segment,
+    segment2: &Segment,
+    bounded: bool,
+) -> Option<Point<SubPixelUnit>> {
+    let p1 = segment1.start;
+    let p2 = segment1.end;
+    let p3 = segment2.start;
+    let p4 = segment2.end;
+    let d = (p4.y - p3.y).mul_add(p2.x - p1.x, -(p4.x - p3.x) * (p2.y - p1.y));
+    if d == 0.0 {
+        return None;
+    }
+    let u_a = (p4.x - p3.x).mul_add(p1.y - p3.y, -(p4.y - p3.y) * (p1.x - p3.x)) / d;
+    let u_b = (p2.x - p1.x).mul_add(p1.y - p3.y, -(p2.y - p1.y) * (p1.x - p3.x)) / d;
+    if !bounded || ((0.0..=1.0).contains(&u_a) && (0.0..=1.0).contains(&u_b)) {
+        return Some(Point::new(
+            u_a.mul_add(p2.x - p1.x, p1.x),
+            u_a.mul_add(p2.y - p1.y, p1.y),
+        ));
+    }
+    None
+}
+
+/// Determines whether the two line segments intersect.
+pub fn segments_intersect(line1: &Segment, line2: &Segment) -> bool {
+    intersection_of_lines(line1, line2, true).is_some()
+}
+
+/// Determines whether a line segment intersects a rectangle.
+pub fn rect_intersects_line(rect: &Rect, line: &Segment) -> bool {
+    let top_left = Point::new(rect.left() as SubPixelUnit, rect.top() as SubPixelUnit);
+    let top_right = Point::new(rect.right() as SubPixelUnit, rect.top() as SubPixelUnit);
+    let bottom_left = Point::new(rect.left() as SubPixelUnit, rect.bottom() as SubPixelUnit);
+    let bottom_right = Point::new(rect.right() as SubPixelUnit, rect.bottom() as SubPixelUnit);
+    let top_line = Segment::new(top_left, top_right);
+    let right_line = Segment::new(top_right, bottom_right);
+    let bottom_line = Segment::new(bottom_left, bottom_right);
+    let left_line = Segment::new(top_left, bottom_left);
+
+    segments_intersect(&top_line, line)
+        || segments_intersect(&bottom_line, line)
+        || segments_intersect(&left_line, line)
+        || segments_intersect(&right_line, line)
+}
+
+/// Returns the angle between two angles in radians.
+pub fn angle_diff(a: Radians, b: Radians) -> Radians {
+    let diff = normalize_angle(a - b);
+    diff.min(PI - diff)
+}
+
+/// Normalize angle to [0, PI). This means that two angles that are
+/// equivalent modulo PI will be equal, e.g. 90° and 270°, even though
+/// they are not equal in the mathematical sense.
+pub fn normalize_angle(angle: Radians) -> Radians {
+    if angle.is_infinite() || angle.is_nan() {
+        return angle;
+    }
+
+    let mut angle = angle % (2.0 * PI);
+    while angle < 0.0 {
+        angle += PI;
+    }
+    while angle >= PI {
+        angle -= PI;
+    }
+    angle
+}
+
+pub fn find_best_line_through_items(
```

Comment:
I had left a comment during initial review that I was having trouble understanding this algorithm. Would still like to walk through it with you, but it's not blocking

Reply:
@jonahkagan and I paired on this. The result is da2bf08.